sdk/src/main/scala/com/horizen/account/sc2sc/AbstractCrossChainRedeemMessageProcessor.scala

```diff
@@ -0,0 +1,181 @@
+package com.horizen.account.sc2sc
+
+import com.horizen.account.events.AddCrossChainRedeemMessage
+import com.horizen.account.sc2sc.AbstractCrossChainRedeemMessageProcessor.Constant.CrossChainMessageHashKey
+import com.horizen.account.state._
+import com.horizen.cryptolibprovider.utils.FieldElementUtils
+import com.horizen.cryptolibprovider.{CryptoLibProvider, Sc2scCircuit}
+import com.horizen.merkletreenative.MerklePath
+import com.horizen.params.NetworkParams
+import com.horizen.sc2sc.{CrossChainMessage, CrossChainMessageHash}
+import com.horizen.utils.BytesUtils
+import sparkz.crypto.hash.Keccak256
+
+trait CrossChainRedeemMessageProvider {
+  def doesScTxCommitmentTreeRootExist(hash: Array[Byte], view: BaseAccountStateView): Boolean
+
+  def doesCrossChainMessageHashFromRedeemMessageExist(hash: CrossChainMessageHash, view: BaseAccountStateView): Boolean
+}
+
+abstract class AbstractCrossChainRedeemMessageProcessor(
+                                                         networkParams: NetworkParams,
+                                                         sc2scCircuit: Sc2scCircuit
+                                                       ) extends NativeSmartContractMsgProcessor with CrossChainRedeemMessageProvider {
+  /**
+   * Apply message to the given view. Possible results:
+   * <ul>
+   * <li>applied as expected: return byte[]</li>
+   * <li>message valid and (partially) executed, but operation "failed": throw ExecutionFailedException</li>
+   * <li>message invalid and must not exist in a block: throw any other Exception</li>
+   * </ul>
+   *
+   * @param msg          message to apply to the state
+   * @param view         state view
+   * @param gas          available gas for the execution
+   * @param blockContext contextual information accessible during execution
+   * @return return data on successful execution
+   */
+  @throws(classOf[ExecutionFailedException])
+  override def process(msg: Message, view: BaseAccountStateView, gas: GasPool, blockContext: BlockContext): Array[Byte] = {
```

Comment:
the process() method is the main entry point of the fake smart contract
inside it we shooud check if the requested function in msg is one defined by our smart contract and redirect to proper action etc...
since we are defining just an abstract class we can leave it unimplemented and delegate subclasses to implement it.
see what has been done in the sender part:
https://github.com/HorizenOfficial/Sidechains-SDK/blob/pc/sc2sc/sdk/src/test/scala/com/horizen/account/sc2sc/CrossChainMessageProcessorFixture.scala#L29

the code here we can be in a protected method processRedeemMesssage(msg: AccountCrossChainRedeemMessage)